### PR TITLE
PyTorch model snippets 

### PR DESCRIPTION
**Pull request by dmatrix**
## What changes are proposed in this pull request?

This is our continued effort to add demonstrative code snippets for the MLflow Model flavor APIs

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ x] `area/docs`: MLflow documentation pages

### How should the PR be classified in the release notes? Choose one:
- [ x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section

### CODE REVIEW

File: mlflow/pytorch/__init__.py [hunk 1]
```diff
@@ -176,51 +252,89 @@ def log_model(
...
-            mlflow.pytorch.log_model(scripted_pytorch_model, "models")
+            mlflow.pytorch.log_model(scripted_pytorch_model, "scripted_models_pth")
+
+        # fetch the logged model artifacts
+        print("--")
+        print("run_id: {}".format(run.info.run_id))
+        for artifact_path in ["models_pth/data", "scripted_models_pth/data"]:
+            artifacts = [f.path for f in MlflowClient().list_artifacts(run.info.run_id,
+                        artifact_path)]
+            print("artifacts: {}".format(artifacts))
```
Comment (harupy) [inline mlflow/pytorch/__init__.py:318]: Maybe just showing these lines should be enough?

Comment (harupy) [reply mlflow/pytorch/__init__.py:318]: We could use `...` too.

```python
class MyModel(str):
    ...


model = MyModel()

# train the model
...

with mlflow.start_run():
    mlflow.pytorch.log_model(model, "model")
```

Comment (dmatrix) [reply mlflow/pytorch/__init__.py:281]: Thanks for the input. For all but `log_model()` and `autolog_log()`,  I have elided the code. However, for the other two methods, I think, it's good to have a full example that they can try.

File: mlflow/pytorch/__init__.py [hunk 2]
```diff
@@ -46,8 +46,47 @@
...
+        ...
+
+        # Log PyTorch model
+        with mlflow.start_run() as run:
+            mlflow.pytorch.log_model(model, "models_pth")
+
+        # Fetch the associated conda environment
+        print("--")
+        print("run_id: {}".format(run.info.run_id))
+        env = mlflow.pytorch.get_default_conda_env()
```
Comment (harupy) [inline mlflow/pytorch/__init__.py:77]: I think just showing `get_default_conda_env` should be enough.

File: mlflow/pytorch/__init__.py [hunk 3]
```diff
@@ -343,24 +418,59 @@ def save_model(
...
-        pytorch_model_path = ...
-        # train our model
-        for epoch in range(500):
-            y_pred = pytorch_model(x_data)
-            ...
-        # Save the model
+
+        # Class defined here
+        class LinearNNModel(torch.nn.Module):
+        ...
```
Comment (harupy) [inline mlflow/pytorch/__init__.py:429]:
```suggestion
        class LinearNNModel(torch.nn.Module):
             ...
```

File: mlflow/pytorch/__init__.py [hunk 4]
```diff
@@ -176,51 +215,87 @@ def log_model(
...
+            print("artifacts: {}".format(artifacts))
+
+    .. code-block:: text
+        :caption: Output
+
+        Epoch: 50.0/250, loss: 0.0003
+        Epoch: 100.0/250, loss: 0.0003
+        Epoch: 150.0/250, loss: 0.0002
+        Epoch: 200.0/250, loss: 0.0002
+        Epoch: 250.0/250, loss: 0.0002
```
Comment (harupy) [inline mlflow/pytorch/__init__.py:288]:
```suggestion
```

I don't think we need to print out the epochs because they are not important information.

File: mlflow/pytorch/__init__.py [hunk 5]
```diff
@@ -343,24 +418,59 @@ def save_model(
...
-            # Saving scripted model
+            # Convert to a scripted model and save it
             scripted_pytorch_model = torch.jit.script(model)
-            mlflow.pytorch.save_model(scripted_pytorch_model, pytorch_model_path)
+            mlflow.pytorch.save_model(scripted_pytorch_model, "scripted_models_pth")
+
+        # Load each saved model for inference
+        print("--")
+        for model_path in ["models_pth", "scripted_models_pth"]:
+            model_uri = "{}/{}".format(os.getcwd(), model_path)
```
Comment (harupy) [inline mlflow/pytorch/__init__.py:449]: Is this line really correct? `"{}/{}".format(os.getcwd(), model_path)` doesn't seem to be the right path.

File: mlflow/pytorch/__init__.py [hunk 6]
```diff
@@ -539,14 +649,44 @@ def load_model(model_uri, **kwargs):
...
+
+        # Training loop
+        ...
+
+        # Log the model
+        with mlflow.start_run() as run:
+            mlflow.pytorch.log_model(model, "models_pth")
+
+        # Inference after loading the logged model
+        print("--")
```
Comment (harupy) [inline mlflow/pytorch/__init__.py:675]:
```suggestion
```

File: mlflow/pytorch/__init__.py [hunk 7]
```diff
@@ -539,14 +649,44 @@ def load_model(model_uri, **kwargs):
     .. code-block:: python
         :caption: Example
 
+        import warnings
+        import os
```
Comment (harupy) [inline mlflow/pytorch/__init__.py:653]:
```suggestion
```

Can we remove these because they don't seem to  be used?

**Comment by harupy** [comment]
Merging the current master branch should fix the issue in the CircleCI build :)

File: mlflow/pytorch/__init__.py (hunk 4 above)
Comment (dmatrix) [reply mlflow/pytorch/__init__.py:288]: I can remove it from the output, but it makes sense to keep it in the code; it's a good indication that your model is merging as the loss is decreasing.

File: mlflow/pytorch/__init__.py (hunk 5 above)
Comment (dmatrix) [reply mlflow/pytorch/__init__.py:449]: Yes, it [seems to work.](https://github.com/dmatrix/mlflow-tests/blob/master/py/mlflow/models/pytorch/save_model.py) The saving model saves it under the local path.

File: mlflow/pytorch/__init__.py (hunk 7 above)
Comment (dmatrix) [reply mlflow/pytorch/__init__.py:653]: As cut-n-paste leftover over from [here](https://github.com/dmatrix/mlflow-tests/blob/master/py/mlflow/models/pytorch/log_model.py). :)

File: mlflow/pytorch/__init__.py (hunk 4 above)
Comment (harupy) [reply mlflow/pytorch/__init__.py]: If this was a tutorial about training a pytorch model, it would make sense to keep the epoch and loss.

Comment (dmatrix) [reply mlflow/pytorch/__init__.py]: I suppose one can always add it if one wants to. Will remove it.

File: mlflow/pytorch/__init__.py [hunk 8]
```diff
@@ -46,8 +46,43 @@
...
+        # Initialize our model, criterion and optimizer
+        model = LinearNNModel()
+        ...
+
+        # Training loop
+        ...
+
+        # Log PyTorch model
+        with mlflow.start_run() as run:
+            mlflow.pytorch.log_model(model, "models_pth")
```
Comment (harupy) [inline mlflow/pytorch/__init__.py:72]:
```suggestion
        import mlflow.pytorch
```

Can we remove unrelated code?

File: mlflow/pytorch/__init__.py (hunk 5 above)
Comment (harupy) [reply mlflow/pytorch/__init__.py:432]: I think this is because the artifact directory points to the current working directory.

File: mlflow/pytorch/__init__.py [hunk 9]
```diff
@@ -539,14 +630,39 @@ def load_model(model_uri, **kwargs):
...
+        ...
+
+        # Log the model
+        with mlflow.start_run() as run:
+            mlflow.pytorch.log_model(model, "models_pth")
+
+        # Inference after loading the logged model
+        model_uri = "runs:/{}/models_pth".format(run.info.run_id)
+        loaded_model = mlflow.pytorch.load_model(model_uri)
+        for hv in [4.0, 6.0, 30.0]:
```
Comment (harupy) [inline mlflow/pytorch/__init__.py:655]: Can we pass `[4.0, 6.0, 30.0]` to the model as a tensor?

```
model(xxx_yyy([4.0, 6.0, 30.0]))
```

File: mlflow/pytorch/__init__.py (hunk 5 above)
Comment (dmatrix) [reply mlflow/pytorch/__init__.py:432]: Is there anything wrong with it If the current artifact directory points to cwd and it's where you wish to save the model?

File: mlflow/pytorch/__init__.py (hunk 8 above)
Comment (dmatrix) [reply mlflow/pytorch/__init__.py:72]: You're minimalist, aren't you? :-) 

Why do you feel that this code is unrelated? This provides some context, so If they want to try a full example, then can
use look at the full example above. At the least, we should have:

        # Class defined here
        class LinearNNModel(torch.nn.Module):
            ...
        # Initialize our model, criterion and optimizer
        ...

        # Training loop
        ...

File: mlflow/pytorch/__init__.py (hunk 9 above)
Comment (dmatrix) [reply mlflow/pytorch/__init__.py:655]: You can do something like this since the network expects the input as shape (n, 1) tensor.

```
X = torch.tensor([4, 6, 30]).view(-1,1).float()
y_pred = loaded_model(X)
print(X)
print(y_pred)
```

But I prefer to simplify and send individual tensor so we can print the X, --> y on the same line. Yes, for larger models
you want your input to be the whole batch, but for here I just made it simple: inferring one tensor at a time.

File: mlflow/pytorch/__init__.py (hunk 5 above)
Comment (harupy) [reply mlflow/pytorch/__init__.py:430]: If the current artifact directory didn't point to cwd, it would fail. I think it's safe to use `mlflow.get_artifact_uri`.

Comment (harupy) [reply mlflow/pytorch/__init__.py:430]: or we could use `runs:{}/{}` format.

File: mlflow/pytorch/__init__.py (hunk 8 above)
Comment (harupy) [reply mlflow/pytorch/__init__.py]: If we had:

```
        # Class defined here
        class LinearNNModel(torch.nn.Module):
            ...
        # Initialize our model, criterion and optimizer
        ...

        # Training loop
        ...
```

it would look like the class definition and training loop affect the result of `get_deafult_conda_env`.

File: mlflow/pytorch/__init__.py [hunk 10]
```diff
@@ -343,24 +401,55 @@ def save_model(
...
-            mlflow.pytorch.save_model(scripted_pytorch_model, pytorch_model_path)
+            mlflow.pytorch.save_model(scripted_pytorch_model, "scripted_models_pth")
+
+        # Load each saved model for inference
+        for model_path in ["models_pth", "scripted_models_pth"]:
+            model_uri = "{}/{}".format(os.getcwd(), model_path)
+            loaded_model = mlflow.pytorch.load_model(model_uri)
+            print("model: {}".format(model_path))
+            for hv in [6.0, 8.0, 12.0, 30.0]:
+                hour_var = torch.Tensor([[hv]])
```
Comment (harupy) [inline mlflow/pytorch/__init__.py:434]: Can we rename `hour_var` to `X`?

File: mlflow/pytorch/__init__.py [hunk 11]
```diff
@@ -343,24 +401,55 @@ def save_model(
...
+                y_pred = loaded_model(hour_var)
+                print("predict X:{}, y_pred: {:.2f}".format(hv, y_pred.data.item()))
```
Comment (harupy) [inline mlflow/pytorch/__init__.py:436]:
```suggestion
                print("predict X: {}, y_pred: {:.2f}".format(hv, y_pred.data.item()))
```

nit

File: mlflow/pytorch/__init__.py (hunk 8 above)
Comment (dmatrix) [reply mlflow/pytorch/__init__.py]: Ah, I see what you mean! 

(The reason I had it is, because if you had just trained and logged a PyTorch model in the elided code above, then you can fetch its conda environment with this API code.)

I can remove it.

File: mlflow/pytorch/__init__.py (hunk 10 above)
Comment (dmatrix) [reply mlflow/pytorch/__init__.py:434]: Yes, indeed!

File: mlflow/pytorch/__init__.py [hunk 12]
```diff
@@ -343,24 +388,55 @@ def save_model(
...
+            mlflow.pytorch.save_model(model, "models_pth")
 
-            # Saving scripted model
+            # Convert to a scripted model and save it
             scripted_pytorch_model = torch.jit.script(model)
-            mlflow.pytorch.save_model(scripted_pytorch_model, pytorch_model_path)
+            mlflow.pytorch.save_model(scripted_pytorch_model, "scripted_models_pth")
+
+        # Load each saved model for inference
+        for model_path in ["models_pth", "scripted_models_pth"]:
```
Comment (harupy) [inline mlflow/pytorch/__init__.py:416]: Why do we suffix directorie with `_pth`?

File: mlflow/pytorch/__init__.py (hunk 5 above)
Comment (harupy) [reply mlflow/pytorch/__init__.py:417]: sorry I misunderstood this is `log_model`.

File: mlflow/pytorch/__init__.py (hunk 12 above)
Comment (dmatrix) [reply mlflow/pytorch/__init__.py:416]: Good observation and point! No need to adopt the PyTorch suffix for model directories. We use generic "models," so better to stick with "models" and scripted_models" respectively.

File: mlflow/pytorch/__init__.py [hunk 13]
```diff
@@ -46,8 +46,30 @@
...
              :func:`save_model()` and :func:`log_model()`.
+
+    .. code-block:: python
+        :caption: Example
+
+        import mlflow.pytorch
+
+        # Log PyTorch model
+        with mlflow.start_run() as run:
+            mlflow.pytorch.log_model(model, "models_pth")
```
Comment (harupy) [inline mlflow/pytorch/__init__.py:59]: Should we remove `_pth` here?

File: mlflow/pytorch/__init__.py [hunk 14]
```diff
@@ -539,14 +614,38 @@ def load_model(model_uri, **kwargs):
...
+            X = torch.Tensor([[x]])
+            y_pred = loaded_model(X)
+            print("predict X: {}, y_pred: {:.2f}".format(x, y_pred.data.item()))
+
+    .. code-block:: text
+        :caption: Output
+
+        predict X:4.0, y_pred: 7.57
+        predict X:6.0, y_pred: 11.64
+        predict X:30.0, y_pred: 60.48
```
Comment (harupy) [inline mlflow/pytorch/__init__.py:648]:
```suggestion
        predict X: 30.0, y_pred: 60.48
```
nit

**Comment by harupy** [comment]
LGTM once https://github.com/mlflow/mlflow/pull/3753#discussion_r538973300 and https://github.com/mlflow/mlflow/pull/3753#discussion_r538973418 have been addressed.

File: mlflow/pytorch/__init__.py (hunk 13 above)
Comment (dmatrix) [reply mlflow/pytorch/__init__.py]: Removed other instances and usage of `_pth` suffix 👍

**Review by harupy — APPROVED** [review]

**Comment by harupy** [comment]
https://github.com/mlflow/mlflow/pull/3802 fixes failures in the R and flavors checks.
